### PR TITLE
Handle non ajax logout request

### PR DESCRIPTION
**Pull request by lightglitch**
| Q | A
| --- | ---
| Bug fix? | yes
| BC breaks? | no

#### What's in this PR?
Handle logout for normal http requests

#### Why?
Sometime we may need to call the logout action from pages that don't use ajax to do it.

### CODE REVIEW

File: src/Sulu/Bundle/SecurityBundle/Resources/config/services.xml [hunk 1]
```diff
@@ -94,15 +94,16 @@
...
+            <argument type="service" id="router"/>
+        </service>
 
         <service id="sulu_security.mask_converter" class="%sulu_security.mask_converter.class%" public="true">
             <argument>%permissions%</argument>
         </service>
 
         <service id="sulu_security.salt_generator" class="%sulu_security.salt_generator.class%" public="true">
         </service>
-        <service id="%sulu_security.salt_generator.class%" alias="sulu_security.salt_generator"/>
```
Comment (danrot) [inline src/Sulu/Bundle/SecurityBundle/Resources/config/services.xml:105]: Was that an accident? I am quite sure we can't simply remove that line.

File: src/Sulu/Bundle/SecurityBundle/Security/LogoutSuccessHandler.php [hunk 2]
```diff
@@ -12,14 +12,29 @@
...
 use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\RedirectResponse;
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\RouterInterface;
 use Symfony\Component\Security\Http\Logout\LogoutSuccessHandlerInterface;
 
 class LogoutSuccessHandler implements LogoutSuccessHandlerInterface
 {
+    private $router;
```
Comment (danrot) [inline src/Sulu/Bundle/SecurityBundle/Security/LogoutSuccessHandler.php:23]: Type that variable using a `@var` annotation:

```php
/**
 * @var RouterInterface
 */
private $router;
```

**Comment by lightglitch** [comment]
@danrot  Both those changes were made when I run cs fixer on those files. Bu I will revert them.

**Comment by danrot** [comment]
And can you also answer what's the use case for this?

File: src/Sulu/Bundle/SecurityBundle/Resources/config/services.xml [hunk 3]
```diff
@@ -94,7 +94,9 @@
             <argument type="service" id="session"/>
         </service>
 
-        <service id="sulu_security.logout_success_handler" class="Sulu\Bundle\SecurityBundle\Security\LogoutSuccessHandler" />
+        <service id="sulu_security.logout_success_handler" class="Sulu\Bundle\SecurityBundle\Security\LogoutSuccessHandler" >
```
Comment (danrot) [inline src/Sulu/Bundle/SecurityBundle/Resources/config/services.xml:97]: No space right before the closing bracket.

**Comment by lightglitch** [comment]
We have an old symfony backoffice that while it's being migrated to sulu "admin framework" still needs to work, so we merge both projects. The authentication system was integrated and in that section of the backoffice they still need to logout without going to sulu interface and it's not using ajax.

**Comment by danrot** [comment]
Ok, and you are reusing the same session to check for login in both applications?

**Comment by lightglitch** [comment]
Yep, same session and also merged with sulu roles and permissions.

**Comment by alexander-schranz** [comment]
Is there any possibility to fallback to the symfony DefaultLogoutSuccessHandler which would take the target configuration into count or any possibility to inject the target configuration?

```yaml
            logout:
                path: sulu_admin.logout
                target: target_route_name
```

**Comment by lightglitch** [comment]
Yep, if LogoutSuccessHandler extends DefaultLogoutSuccessHandler.

**Comment by danrot** [comment]
If we are doing this I would prefer to [decorate the service](https://symfony.com/doc/current/service_container/service_decoration.html). Then the configuration etc. is already correctly passed by Symfony itself.

**Comment by lightglitch** [comment]
Well symfony only passes the target configuration if the handler is not defined:
```
            // add logout success handler
            if (isset($firewall['logout']['success_handler'])) {
                $logoutSuccessHandlerId = $firewall['logout']['success_handler'];
            } else {
                $logoutSuccessHandlerId = 'security.logout.success_handler.'.$id;
                $logoutSuccessHandler = $container->setDefinition($logoutSuccessHandlerId, new ChildDefinition('security.logout.success_handler'));
                $logoutSuccessHandler->replaceArgument(1, $firewall['logout']['target']);
            }
```
So we can't get target value for our listener.

**Comment by nnatter** [comment]
If we use decoration for the `LogoutSuccessHandlerInterface`, shouldnt we do the same thing for the `AuthenticationSuccessHandlerInterface` and the `AuthenticationFailureHandlerInterface`? 

Both of them are currently implemented by the `AuthenticationHandler` and use a hardcoded default behaviour like implemented in this PR.

**Comment by danrot** [comment]
> Well symfony only passes the target configuration if the handler is not defined:
> 
> ```
>             // add logout success handler
>             if (isset($firewall['logout']['success_handler'])) {
>                 $logoutSuccessHandlerId = $firewall['logout']['success_handler'];
>             } else {
>                 $logoutSuccessHandlerId = 'security.logout.success_handler.'.$id;
>                 $logoutSuccessHandler = $container->setDefinition($logoutSuccessHandlerId, new ChildDefinition('security.logout.success_handler'));
>                 $logoutSuccessHandler->replaceArgument(1, $firewall['logout']['target']);
>             }
> ```
> 
> So we can't get target value for our listener.

That's some weird code :confused: But it looks to me like this would still work, if you don't pass and `success_handler` via the config, but just decorate the existing service instead.

**Comment by danrot** [comment]
> If we use decoration for the `LogoutSuccessHandlerInterface`, shouldnt we do the same thing for the `AuthenticationSuccessHandlerInterface` and the `AuthenticationFailureHandlerInterface`?
> 
> Both of them are currently implemented by the `AuthenticationHandler` and use a hardcoded default behaviour like implemented in this PR.

If we need the behavior of the other handler as well I would say ideally yes.

**Comment by nnatter** [comment]
> If we need the behavior of the other handler as well I would say ideally yes.
Not sure if we need them. At leas until now, we did not 😅 Guess my point is that it would be okay for me to hardcode the behaviour in the `LogoutSuccessHandler` too, because we already did that in the `AuthenticationHandler`.

**Comment by danrot** [comment]
Yeah, since the behavior is better than before this is ok for me as well :see_no_evil: If @alexander-schranz can live with that I am fine with merging that.

@lightglitch But at least you should rebase the PR once more, I have fixed the `php-cs-fixer` issue that breaks your build in another PR.

**Comment by alexander-schranz** [comment]
@danrot is fine for me.

**Comment by lightglitch** [comment]
@danrot done.

**Comment by danrot** [comment]
@lightglitch Thank you for your contribution!

**Pull request merged**